### ISSUE TITLE
Use cases

### ISSUE THREAD

**Issue by kelunik** [fiberphp/fiber-ext#21]
Let's collect use cases fibers can be used for. The one we (@amphp) have is non-blocking I/O and timers. Are there any others? What do they `yield` and how are they continued? What does the scheduler do with the yielded value?

**Comment by uasan** [comment]
Basically, it is IO, and events (timers ...) but there is still such use case - `defer` as in the golang
https://tour.golang.org/flowcontrol/12

Now use the Generators, when the Fiber is stabilized try to rewrite it on the Fiber.

**Comment by kelunik** [comment]
Defer can already be implemented by abusing `__destruct()` and isn't really related to fibers, but rather a separate feature that could be added to PHP, see https://preprocess.io/#deferred.

If event loops are the only use case, we can merge something like [`green-thread`](https://github.com/amphp/green-thread/blob/83dc57e80d48e49ade0c5224e0e3f078f87e1775/src/functions.php) directly into this extension. `Fiber` would implement `Promise` then. Anything that needs the event loop would create a `Deferred`, pass that to the event loop and resolve it in the event loop callback. The outer code just awaits the `Promise` of the `Deferred` then. Our fiber based implementation of `amphp/byte-stream` currently [does exactly that](https://github.com/amphp/byte-stream/blob/509737552b3e2ef31defc08862eea32adcd7d92a/lib/ResourceInputStream.php#L48-L95).

This avoids an unnecessary wrapping of fibers if anything that's going to use that is related to event loops and promises anyway.

**Comment by uasan** [comment]
> Defer can already be implemented by abusing __destruct() 

Yes, but `defer`, there must be the lowest execution priority, in fact, it is executed when the event loop is already empty.


> This avoids an unnecessary wrapping of fibers if anything that's going to use that is related to event loops and promises anyway.

I think, Fiber, need a normal application server (not `php-cli`).
The [Nginx Unit](http://unit.nginx.org/) is the most perspective and I hope someone will make module PHP for Nginx Unit as application server for PHP script.

**Comment by kelunik** [comment]
As far as I know `defer` would be executed as soon as the function it is called is left. It's not what `Amp\Loop::defer()` does.

What do you mean by "normal application server"?

**Comment by uasan** [comment]
> What do you mean by "normal application server"? Why not php-cli?

A single point on the management of microservices on different technologies, PHP, Python, Node.js, Golang...

Internals HTTP server and cache HTTP responses.

Now every async PHP Frameworks, makes its HTTP parser for its HTTP server.

In the Nginx Unit, this is all out of the box, it already has a module for PHP similar to PHP-FPM.

But need a new module Nginx Unit, for async PHP which will have a built-in event loop (maybe libuv) and HTTP Server, and cache HTTP responses...

**Comment by kelunik** [comment]
We already have that with https://github.com/amphp/aerys and https://github.com/reactphp/http. Nginx Unit will just be an additional third implementation. With my proposal outlined in #22 the event loop could be totally hidden from PHP users without the need for an event loop built in.

**Comment by uasan** [comment]
> We already have that with https://github.com/amphp/aerys and https://github.com/reactphp/http. Nginx Unit will just be an additional third implementation. 

No, Nginx Unit must be used in `aerys` and `reactphp` how PHP-FPM is used in Symphony, Zend Framework, and hundreds of other php frameworks.

Then there one implementation of the HTTP server and the event loop, on SAPI PHP levels, not at the PHP userland frameworks.

**Comment by kelunik** [comment]
It doesn't matter where it is implemented, people just need to agree on one implementation or build adapters. The only reason to have that in core would be extensions being able to use it.

Nginx Unit is specific to HTTP services, but that's just a subset of all applications making use of an event loop.

**Comment by uasan** [comment]
> It doesn't matter where it is implemented, people just need to agree on one implementation or build adapters. The only reason to have that in core would be extensions being able to use it.

It is important, for two reasons, the technical aspect and marketing :)

1. Technical aspect, SAPI will spend less memory and resources on socket processing than PHP in userland, SAPI will only call a callback function from userland and send the received data to it, SAPI implements HTTP cache response, without call PHP...

2. I can not sell to my management, to `aerys `and `reactphp`, but to sell Nginx Unit, I can this brand trust him... :)

**Comment by lvht** [comment]
The https://github.com/fiberphp/fiber-demo is another usage example.

**Comment by martinschroeder** [comment]
A very nice use case would be to create a PSR-7 stream adapter (something like https://github.com/martinschroeder/fiber-test/blob/master/src/InputStreamAdapter.php) that can be used to provide access to HTTP server request bodies using https://github.com/amphp/http-server. We could use https://github.com/zendframework/zend-diactoros for PSR-7 classes and just add that little adapter.

This approach could be used to show how to use an arbitrary PHP framework (that supports PSR-7) to serve multiple requests from a single PHP process (a minimal example of what something like https://github.com/php-pm/php-pm can do).

Of course it is also possible to do the same thing for an HTTP client (like already implemented in https://github.com/fiberphp/fiber-http) to show that multiple concurrent HTTP requests are possible without exposing something like a promise-bases async API to consumers (you just pass an HTTP request and the client will return an HTTP response).

**Comment by joelwurtz** [comment]
For me, example with PSR7 is really the killer feature because it shows that fiber allow doing async with existing interface / specifications withtout breaking the contract and so with this, async becomes only an implementation problem not a contract one like actually (as we need promise or event based system)

**Comment by martinschroeder** [comment]
I have created a little [demo repository](https://github.com/martinschroeder/fiber-demo) that combines an async HTTP server (Amp) with `PSR-7` messages and `PSR-15` middlewares. I added some features like routing (FastRoute) to create a demo application that looks and feels like a typical microframework. This approach can be used to compose PSR-based components into an async application.

It is not the same thing as running a framework like Symfony or laravel yet, but it shows that it can be done. The real problem with bigger frameworks will be that they still use sync IO when dealing with the filesystem or a database that will block the event loop and slow things down. Anyways i think it is very promising and quit fast. ;-)

**Comment by yannsun** [comment]
For async php framework, I recommend to try and study the [swoole-src](https://github.com/swoole/swoole-src) project. It has been widely applied in my company and it provides more features than [fiber](https://github.com/fiberphp/fiber-ext), for example async IO which is based on AIO and coroutine which is released in [swoole2.0](https://github.com/swoole/swoole-src/blob/master/Version2.md) and it is a more proven technique than fiber.  As for combining with PSR standards, I recommend to try [swoft](https://github.com/swoft-cloud/swoft) which is based on [Swoole](https://github.com/swoole/swoole-src) . Except for implement PSR standard, it has done a lot of work to make the job easier to build a high-performance systems.
